airflow/operators/branch.py

```diff
@@ -34,8 +36,42 @@ class BranchMixIn(SkipMixin):
     def do_branch(self, context: Context, branches_to_execute: str | Iterable[str]) -> str | Iterable[str]:
         """Implement the handling of branching including logging."""
         self.log.info("Branch into %s", branches_to_execute)
-        self.skip_all_except(context["ti"], branches_to_execute)
-        return branches_to_execute
+        branch_task_ids = self.expand_task_group_roots(context["ti"], branches_to_execute)
+        self.skip_all_except(context["ti"], branch_task_ids)
+        return branch_task_ids
+
+    def expand_task_group_roots(
+        self, ti: TaskInstance | TaskInstancePydantic, branches_to_execute: str | Iterable[str]
+    ) -> str | Iterable[str]:
```

Comment:
Does this function need to be public? We should add an underscore prefix unless necessary.

Also it would be better if this is a generator that yields task IDs instead.

Reply:
It does not have to be public, I will prefix it with underscore.

If I change this to a generator, how should I modify this part?
https://github.com/apache/airflow/blob/e69853ac9e222259bb9963b3b8359c294d84e148/airflow/operators/branch.py#L39

Should I convert it into list from the generator something like this?
```python
branch_task_ids = list(self._expand_task_group_roots(context["ti"], branches_to_execute))

Reply:
You don’t need to change the call site; `skip_all_except` supports passing in an iterator.

Reply:
Not sure if I need to change that, `skip_all_except` already accepts passing in an iterator and with the latest commit with generator, it looks like it works fine. What do you think?

Reply:
It accepts any iterator but does not handle it correctly. `branch_task_ids` is iterated twice in the function, which you can’t do with an iterator. The line above is the second iteration we need to get rid of.

Reply:
It appears to “work fine” because the generator quietly yields nothing after it is exhausted, and the line becomes a no-op.

Reply:
Oh, got it now, thanks! Changed it as you recommended.